README.md
README
================
Francisco Bischoff
\- 12 Mar 2020

<!-- README.md is generated from README.Rmd. Please edit that file -->

<img src="man/figures/logo.png" align="right" style="float:right;" />

# Time Series with Matrix Profile

<!-- badges: start -->
[![Packagist](https://img.shields.io/badge/License-Apache--2.0-brightgreen.svg)](https://choosealicense.com/licenses/apache-2.0/)
[![lifecycle](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://www.tidyverse.org/lifecycle/#stable)
[![CRAN
version](http://www.r-pkg.org/badges/version/tsmp)](https://cran.r-project.org/package=tsmp)
[![CRAN
Downloads](https://cranlogs.r-pkg.org/badges/tsmp)](https://cran.r-project.org/package=tsmp)
[![CircleCI build
status](https://circleci.com/gh/matrix-profile-foundation/tsmp.svg?style=svg)](https://circleci.com/gh/matrix-profile-foundation/tsmp)
<!-- badges: end -->

|               | Build                                                                                                                                                                             | Dev                                                                                                                                                                                 |
| ------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Linux x86\_64 | [![Build Status](https://travis-ci.com/matrix-profile-foundation/tsmp.svg?branch=master)](https://travis-ci.com/matrix-profile-foundation/tsmp)                                   | [![Build Status](https://travis-ci.com/matrix-profile-foundation/tsmp.svg?branch=develop)](https://travis-ci.com/matrix-profile-foundation/tsmp)                                    |
| OSX           | [![Build Status](https://travis-ci.com/matrix-profile-foundation/tsmp.svg?branch=master)](https://travis-ci.com/matrix-profile-foundation/tsmp)                                   | [![Build Status](https://travis-ci.com/matrix-profile-foundation/tsmp.svg?branch=develop)](https://travis-ci.com/matrix-profile-foundation/tsmp)                                    |
| Windows       | [![AppVeyor build status](https://ci.appveyor.com/api/projects/status/byfyqncr60ten98g/branch/master?svg=true)](https://ci.appveyor.com/project/franzbischoff/tsmp/branch/master) | [![AppVeyor build status](https://ci.appveyor.com/api/projects/status/byfyqncr60ten98g/branch/develop?svg=true)](https://ci.appveyor.com/project/franzbischoff/tsmp/branch/develop) |
| Coverage      | [![codecov](https://codecov.io/gh/matrix-profile-foundation/tsmp/branch/master/graph/badge.svg)](https://codecov.io/gh/matrix-profile-foundation/tsmp)                            | [![codecov](https://codecov.io/gh/matrix-profile-foundation/tsmp/branch/develop/graph/badge.svg)](https://codecov.io/gh/matrix-profile-foundation/tsmp)                             |

## Overview

R Functions implementing UCR Matrix Profile Algorithm
(<http://www.cs.ucr.edu/~eamonn/MatrixProfile.html>).

This package allows you to use the Matrix Profile concept as a toolkit.

This package provides:

  - Algorithms to build a Matrix Profile: STAMP, STOMP, SCRIMP++,
    SIMPLE, MSTOMP and VALMOD.
  - Algorithms for MOTIF search for Unidimensional and Multidimensional
    Matrix Profiles.
  - Algorithm for Chains search for Unidimensional Matrix Profile.
  - Algorithms for Semantic Segmentation (FLUSS) and Weakly Labeled data
    (SDTS).
  - Algorithm for Salient Subsections detection allowing MDS plotting.
  - Basic plotting for all outputs generated here.
  - Sequencial workflow, see below.

<!-- end list -->

``` r
# Basic workflow:
matrix <- tsmp(data, window_size = 30) %>%
  find_motif(n_motifs = 3) %T>%
  plot()

# SDTS still have a unique way to work:
model <- sdts_train(data, labels, windows)
result <- sdts_predict(model, data, round(mean(windows)))
```

Please refer to the [User
Manual](https://matrix-profile-foundation.github.io/tsmp/reference/) for
more details.

Please be welcome to suggest improvements.

### Performance on an Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz using a random walk dataset

``` r
set.seed(2018)
data <- cumsum(sample(c(-1, 1), 40000, TRUE))
```

#### Current version benchmark

|             | Elapsed Time(s) | Data Size | Window Size | Threads | Lang |
| ----------- | --------------: | --------: | ----------: | ------: | :--- |
| `mpx_par`   |            0.59 |     40000 |        1000 |       8 | Rcpp |
| `mpx`       |            1.94 |     40000 |        1000 |       1 | Rcpp |
| `stomp_par` |           38.90 |     40000 |        1000 |       8 | R    |
| `stomp`     |           85.13 |     40000 |        1000 |       1 | R    |
| `scrimp`    |          123.07 |     40000 |        1000 |       1 | R    |
| `stamp_par` |          925.45 |     40000 |        1000 |       8 | R    |
| `stamp`     |         3776.86 |     40000 |        1000 |       1 | R    |

## Installation

``` r
# Install the released version from CRAN
install.packages("tsmp")

# Or the development version from GitHub:
# install.packages("devtools")
devtools::install_github("matrix-profile-foundation/tsmp")
```

## Currently available Features

  - STAMP (single and multi-thread versions)
  - STOMP (single and multi-thread versions)
  - STOMPi (On-line version)
  - SCRIMP (single-thread, not for AB-joins yet)
  - Time Series Chains
  - Multivariate STOMP (mSTOMP)
  - Multivariate MOTIF Search (from mSTOMP)
  - Salient Subsequences search for Multidimensional Space
  - Scalable Dictionary learning for Time Series (SDTS) prediction
  - FLUSS (Fast Low-cost Unipotent Semantic Segmentation)
  - FLOSS (Fast Low-cost On-line Unipotent Semantic Segmentation)
  - SiMPle-Fast (Fast Similarity Matrix Profile for Music Analysis and
    Exploration)
  - Annotation vectors (e.g., Stop-word MOTIF bias, Actionability bias)
  - FLUSS Arc Plot and SiMPle Arc Plot
  - Exact Detection of Variable Length Motifs (VALMOD)
  - MPdist: Matrix Profile Distance
  - Time Series Snippets
  - Subsetting Matrix Profiles (`head()`, `tail()`, `[`, etc.)
  - Misc:
      - MASS v2.0
      - MASS v3.0
      - MASS extensions: ADP (Approximate Distance Profile, with PAA)
      - MASS extensions: WQ (Weighted Query)
      - MASS extensions: QwG (Query with Gap)
      - Fast moving average
      - Fast moving SD

## Roadmap

  - Profile-Based Shapelet Discovery
  - GPU-STOMP

## Other projects with Matrix Profile

  - Python: <https://github.com/target/matrixprofile-ts>
  - Python: <https://github.com/ZiyaoWei/pyMatrixProfile>
  - Python: <https://github.com/jbeleno/owlpy>
  - Python: <https://github.com/javidlakha/matrix-profile>
  - Python: <https://github.com/shapelets/khiva-python>
  - R: <https://github.com/shapelets/khiva-r>
  - Matlab: <https://github.com/shapelets/khiva-matlab>
  - Java: <https://github.com/shapelets/khiva-java>
  - Java: <https://github.com/ensozos/Matrix-Profile>
  - Kotlin: <https://github.com/shapelets/khiva-kotlin>
  - C++ (CUDA and OPENCL): <https://github.com/shapelets/khiva>
  - CUDA: <https://github.com/zpzim/STOMPSelfJoin>
  - CUDA: <https://github.com/zpzim/SCAMP>

## Matrix Profile Foundation

Our next step unifying the Matrix Profile implementation in several
programming languages.

Visit: [Matrix Profile Foundation](https://matrixprofile.org)

## Package dependencies

![](man/figures/dependency_plot-1.png)<!-- -->

## Code of Conduct

Please note that the ‘tsmp’ project is released with a [Contributor Code
of
Conduct](https://github.com/matrix-profile-foundation/tsmp/blob/master/.github/CODE_OF_CONDUCT.md).
By contributing to this project, you agree to abide by its terms.


NEWS
<!-- NEWS.md is generated from NEWS.Rmd. Please edit that file -->

tsmp 0.4.11
===========

-   Fix in `mpx()`.

tsmp 0.4.9
==========

-   Fixed MacOSX build.

tsmp 0.4.8
==========

-   Added MPdist algorithm to compare two time series.
-   Added `find_snippet()` that uses MPdist to show representative data.
-   Added `mpx()` algorithm that doesn’t depends on FFT.
-   Added `pmp()` pan-matrix profile.
-   Improvement with several implementations in Rcpp. Computation speed
    is much faster.
-   Added `compute()`, `analyze()` and `visualize()`. Starting point in
    the unified API from MPF.

tsmp 0.3.5
==========

-   Changed `mass()` to `dist_profile()`, including options to different
    algorithms.
-   Added MASS\_V3 and MASS\_Weighted to `dist_profile()` algorithms.
-   Function `dist_profile()` allows Query with Gap (QwG) and
    Approximate Distance Profile (ADP, with PAA)
-   Fixed long runtime of FFT for some data sizes, using MASS\_V3. Issue
    \#36.
-   Added `stompi_update()` that updates the matrix profile allowing
    real-time computation.
-   Added `floss()` which can do real-time FLUSS computation.
-   Added subset operator `[` for `tsmp` objects.
-   Added `tail()` and `head()` for `tsmp` objects.
-   Changed `find_motif()` for `MultiMatrixProfile` to report Motifs
    correctly.

tsmp 0.3.4
==========

-   Added `find_discord()` and its `print()` and `plot()` functions.
-   Changed `plot()` for motifs to show where are the neighbors. Same
    for discord.
-   Added `valmod()` for Variable Length Motif Discovery.
-   Changed `find_motif()` for compatibility with `valmod()`.

tsmp 0.3.3
==========

-   Fixed `find_chains()` not returning the longest chain. Issue \#33

tsmp 0.3.2
==========

-   Fixed Matrix Profile print, dimensions are now reported correctly.
-   Fixed pipe imports. Issue \#22
-   Fixed bug with `vars`. Issue \#23
-   Changed package license to GPL-3.
-   Changed verbose mode, added one more step to separate messages from
    progression bar.
-   Fixed SCRIMP and added PRE-SCRIMP, so this is the SCRIMP++. AB-join
    not yet implemented.
-   Changed progress bar for a better one from `progress` package.
-   Added Print and Plot to SiMPle. Issue \#24
-   Added Print and Plot to Salient.

tsmp 0.3.1
==========

IMPORTANT
---------

-   This version is a complete restructuration. The API has changed, and
    the workflow is more friendly. This API is intended to be stable,
    and from now on any change will pass through the “Deprecated” stage.

Added Features
--------------

-   Outputs have a prettier print format.
-   Outputs have a plot function. Try to plot a `tsmp()` output for
    example.
-   Now functions can work in `%>%` (pipe),
    e.g. `tsmp() %>% find_motif()`. Except for SDTS that has a proper
    way to work.
-   Added a wrapper function called `tsmp()` that handles the several
    algorithms available.
-   Added `as.*` functions to allow you to switch classes if you want,
    e.g.: `as.matrixprofile()`.
-   Changed all functions from dotted.case to snake\_case (except `as.*`
    functions).
-   Added Annotation Vectors.
-   Fixed STOMP crash with Joins.
-   Added support to query &lt; data in Joins.
-   SCRIMP (experimental).

tsmp 0.2.15
===========

-   Code linting.
-   Added Salient Subsequences search.

tsmp 0.2.14
===========

-   Added SiMPle (Fast Similarity Matrix Profile for Music Analysis and
    Exploration).
-   Added FLUSS (Fast Low-cost Unipotent Semantic Segmentation).
-   Added \[find\_chains()\] to look for chains primitives.
-   Added Multivariate MOTIF Search (from mSTOMP)
-   Changed dependency from beepr to audio (actually beepr depends on
    audio, so fewer dependencies).
-   Added a `NEWS.md` file to track changes to the package.

tsmp 0.2.12
===========

-   Added Multivariate STOMP parallel version.
-   Added SDTS algorithm (Scalable Dictionary learning for Time Series).

tsmp 0.1.0
==========

-   STAMP and STAMP parallel Algorithm.
-   Multivariate STOMP algorithm.
-   MASS algorithm.


R/tsmp.R
#' Computes the Matrix Profile and Profile Index
#'
#' This is a wrap function that makes easy to use all available algorithms to compute the Matrix
#' Profile and Profile Index for multiple purposes.
#'
#' @details The Matrix Profile, has the potential to revolutionize time series data mining because
#'   of its generality, versatility, simplicity and scalability. In particular it has implications
#'   for time series motif discovery, time series joins, shapelet discovery (classification),
#'   density estimation, semantic segmentation, visualization, rule discovery, clustering etc.
#'
#'   The first algorithm invented was the [stamp()] that using [mass()] as an ultra-fast Algorithm
#'   for Similarity Search allowed to compute the Matrix Profile in reasonable time. One of its main
#'   feature was its Anytime property which using a randomized approach could return a "best-so-far"
#'   matrix that could give us the correct answer (using for example 1/10 of all iterations) almost
#'   every time.
#'
#'   The next algorithm was [stomp()] that currently is the most used. Researchers noticed that the
#'   dot products do not need to be recalculated from scratch for each subsequence. Instead, we can
#'   reuse the values calculated for the first subsequence to make a faster calculation in the next
#'   iterations. The idea is to make use of the intersections between the required products in
#'   consecutive iterations. This approach reduced the time to compute the Matrix Profile to about
#'   3% compared to [stamp()], but on the other hand, we lost the Anytime property.
#'
#'   Currently there is a new algorithm that I'll not explain further here. It is called [scrimp()],
#'   and is as fast as [stomp()], and have the Anytime property. This algorithm is implemented in
#'   this package, but still waiting for an article publication.
#'
#'   Further, there is the [mstomp()] that computes a multidimensional Matrix Profile that allows to
#'   meaningful MOTIF discovery in Multivariate Time Series. And [simple_fast()] that also handles
#'   Multivariate Time Series, but focused in Music Analysis and Exploration.
#'
#'   The [valmod()] uses a new pruning algorithm allowing a similarity search with a range of sliding
#'   window sizes.
#'
#'   Some parameters are global across the algorithms:
#'   \describe{
#'     \item{...}{One or two time series (except for [mstomp()]). The second time series can be smaller than the first.}
#'     \item{window_size}{The sliding window.}
#'     \item{exclusion_zone}{Is used to avoid trivial matches; if a query data is provided
#'       (join similarity), this parameter is ignored.}
#'     \item{verbose}{Changes how much information is printed by this function; `0` means nothing,
#'     `1` means text, `2` adds the progress bar, `3` adds the finish sound.}
#'     \item{n_workers}{number of threads for parallel computing (except `simple_fast`, `scrimp` and `valmod`).
#'     If the value is 2 or more, the '_par' version of the algorithm will be used.}
#'   }
#'
#'   `s_size` is used only in Anytime algorithms: [stamp()] and [scrimp()].
#'   `must_dim` and `exc_dim` are used only in [mstomp()].
#'   `heap_size` is used only for [valmod()]
#'   `mode` can be any of the following: `stomp`, `stamp`, `simple`, `mstomp`, `scrimp`, `valmod`.
#'
#' @param ... a `matrix` or a `vector`. If a second time series is supplied it will be a join matrix
#'   profile (except for [mstomp()]).
#' @param window_size an `int` with the size of the sliding window. Use a vector for Valmod.
#' @param exclusion_zone a `numeric`. Size of the exclusion zone, based on window size (default is
#'   `1/2`). See details.
#' @param verbose an `int`. (Default is `2`). See details.
#' @param n_workers an `int`. Number of workers for parallel. (Default is `1`).
#' @param mode the algorithm that will be used to compute the matrix profile. (Default is `stomp`).
#'   See details.
#' @param s_size a `numeric`. for anytime algorithm, represents the size (in observations) the
#'   random calculation will occur (default is `Inf`). See details.
#' @param must_dim an `int` or `vector` of which dimensions to forcibly include (default is `NULL`).
#'   See details.
#' @param exc_dim an `int` or `vector` of which dimensions to exclude (default is `NULL`). See
#'   details.
#' @param heap_size an `int`. (Default is `50`). Size of the distance profile heap buffer.
#' @param .keep_data a `logical`. (Default is `TRUE`). Keeps the data embedded to resultant object.
#'
#' @return Returns the matrix profile `mp` and profile index `pi`. It also returns the left and
#'   right matrix profile `lmp`, `rmp` and profile index `lpi`, `rpi` that may be used to detect
#'   Time Series Chains. [mstomp()] returns a multidimensional Matrix Profile.
#' @export
#' @references * Silva D, Yeh C, Batista G, Keogh E. Simple: Assessing Music Similarity Using
#'   Subsequences Joins. Proc 17th ISMIR Conf. 2016;23–30.
#' @references * Silva DF, Yeh C-CM, Zhu Y, Batista G, Keogh E. Fast Similarity Matrix Profile for
#'   Music Analysis and Exploration. IEEE Trans Multimed. 2018;14(8):1–1.
#' @references * Yeh CM, Kavantzas N, Keogh E. Matrix Profile VI : Meaningful Multidimensional Motif
#'   Discovery.
#' @references * Yeh CCM, Zhu Y, Ulanova L, Begum N, Ding Y, Dau HA, et al. Matrix profile I: All
#'   pairs similarity joins for time series: A unifying view that includes motifs, discords and
#'   shapelets. Proc - IEEE Int Conf Data Mining, ICDM. 2017;1317–22.
#' @references * Zhu Y, Imamura M, Nikovski D, Keogh E. Matrix Profile VII: Time Series Chains: A
#'   New Primitive for Time Series Data Mining. Knowl Inf Syst. 2018 Jun 2;1–27.
#' @references * Zhu Y, Zimmerman Z, Senobari NS, Yeh CM, Funning G. Matrix Profile II : Exploiting
#'   a Novel Algorithm and GPUs to Break the One Hundred Million Barrier for Time Series Motifs and
#'   Joins. Icdm. 2016 Jan 22;54(1):739–48.
#' @references Website: <https://sites.google.com/view/simple-fast>
#' @references Website: <https://sites.google.com/site/ismir2016simple/home>
#' @references Website: <http://www.cs.ucr.edu/~eamonn/MatrixProfile.html>
#' @family matrix profile computations
#' @examples
#' # default with [stomp()]
#' mp <- tsmp(mp_toy_data$data[1:200, 1], window_size = 30, verbose = 0)
#' 
#' # parallel with [stomp_par()]
#' mp <- tsmp(mp_toy_data$data[1:200, 1], window_size = 30, n_workers = 2, verbose = 0)
#' 
#' # Anytime STAMP
#' mp <- tsmp(mp_toy_data$data[1:200, 1], window_size = 30, mode = "stamp", s_size = 50, verbose = 0)
#' 
#' # [mstomp()]
#' mp <- tsmp(mp_toy_data$data[1:200, ], window_size = 30, mode = "mstomp", verbose = 0)
#' 
#' # [simple_fast()]
#' mp <- tsmp(mp_toy_data$data[1:200, ], window_size = 30, mode = "simple", verbose = 0)
tsmp <- function(..., window_size, exclusion_zone = 1 / 2, mode = c("stomp", "stamp", "simple", "mstomp", "scrimp", "valmod"),
                 verbose = 2, n_workers = 1, s_size = Inf, must_dim = NULL, exc_dim = NULL, heap_size = 50, .keep_data = TRUE) {
  algo <- match.arg(mode)

  if (length(list(...)) == 0) {
    stop("Error: You must supply at least one time series.")
  }

  result <- switch(algo,
    "stomp" = {
      if (length(list(...)) > 2) {
        warning("Warning: Only the first two time series will be used.")
      }
      if (n_workers > 1) {
        stomp_par(...,
          window_size = min(window_size), exclusion_zone = exclusion_zone,
          verbose = verbose, n_workers = n_workers
        )
      } else {
        stomp(...,
          window_size = min(window_size), exclusion_zone = exclusion_zone,
          verbose = verbose
        )
      }
    },
    "stamp" = {
      if (length(list(...)) > 2) {
        warning("Warning: Only the first two time series will be used.")
      }

      if (n_workers > 1) {
        stamp_par(...,
          window_size = min(window_size), exclusion_zone = exclusion_zone,
          verbose = verbose, s_size = s_size, n_workers = n_workers
        )
      } else {
        stamp(...,
          window_size = min(window_size), exclusion_zone = exclusion_zone,
          verbose = verbose, s_size = s_size
        )
      }
    },
    "simple" = {
      if (length(list(...)) > 2) {
        warning("Warning: Only the first two time series will be used.")
      }
      simple_fast(...,
        window_size = min(window_size), exclusion_zone = exclusion_zone,
        verbose = verbose
      )
    },
    "mstomp" = {
      if (length(list(...)) > 1) {
        warning("Warning: Only the first time series will be used in `mstomp`.")
      }

      if (n_workers > 1) {
        mstomp_par(list(...)[[1]],
          window_size = min(window_size), exclusion_zone = exclusion_zone,
          verbose = verbose, must_dim = must_dim, exc_dim = exc_dim, n_workers = n_workers
        )
      } else {
        mstomp(list(...)[[1]],
          window_size = min(window_size), exclusion_zone = exclusion_zone,
          verbose = verbose, must_dim = must_dim, exc_dim = exc_dim
        )
      }
    },
    "scrimp" = {
      if (length(list(...)) > 2) {
        warning("Warning: Only the first two time series will be used.")
      }

      scrimp(...,
        window_size = min(window_size), exclusion_zone = exclusion_zone,
        verbose = verbose, s_size = s_size
      )
    },
    "valmod" = {
      if (length(list(...)) > 2) {
        warning("Warning: Only the first two time series will be used.")
      }

      valmod(...,
        window_min = min(window_size), window_max = max(window_size), heap_size = heap_size, exclusion_zone = exclusion_zone,
        verbose = verbose
      )
    },
    stop("Error: `mode` must be ", mode)
  )

  if (.keep_data) {
    args <- list(...)
    result$data <- lapply(args, as.matrix)
  }

  return(result)
}
